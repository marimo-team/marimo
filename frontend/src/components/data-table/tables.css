.table-container {
  @apply rounded-md border overflow-hidden;
}
