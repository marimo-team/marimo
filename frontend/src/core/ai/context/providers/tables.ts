/* Copyright 2024 Marimo. All rights reserved. */

import type { Completion } from "@codemirror/autocomplete";
import type { DatasetTablesMap } from "@/core/datasets/data-source-connections";
import type { DataTable } from "@/core/kernel/messages";
import type { AIContextItem } from "../registry";
import { AIContextProvider } from "../registry";
import { Boosts } from "./common";

export interface TableContextItem extends AIContextItem {
  type: "data";
  data: DataTable;
}

export class TableContextProvider extends AIContextProvider<TableContextItem> {
  readonly title = "Tables";
  readonly mentionPrefix = "@";
  readonly contextType = "data";

  constructor(private tablesMap: DatasetTablesMap) {
    super();
  }

  getItems(): TableContextItem[] {
    return [...this.tablesMap.entries()].map(([tableName, table]) => ({
      uri: this.asURI(tableName),
      name: tableName,
      type: this.contextType,
      description: table.source === "memory" ? "in-memory" : table.source,
      data: table,
    }));
  }

  formatContext(item: TableContextItem): string {
    const { data, uri: id } = item;
    const { columns, source, num_rows, num_columns } = data;
    const shape = [
      num_rows == null ? undefined : `${num_rows} rows`,
      num_columns == null ? undefined : `${num_columns} columns`,
    ]
      .filter(Boolean)
      .join(", ");

    let context = `Table: ${id}\nSource: ${source || "unknown"}`;
    if (shape) {
      context += `\nShape: ${shape}`;
    }

    if (columns && columns.length > 0) {
      context += `\nColumns:\n${columns.map((col) => `  - ${col.name}: ${col.type}`).join("\n")}`;
    }

    return context;
  }

  getCompletions(): Completion[] {
    return [];
  }

  formatCompletion(item: TableContextItem): Completion {
    const tableName = item.data.name;
    const table = item.data;
    return {
      label: `@${tableName}`,
      displayLabel: tableName,
      detail: table.source === "memory" ? "in-memory" : table.source,
      boost:
        table.source_type === "local"
          ? Boosts.LOCAL_TABLE
          : Boosts.REMOTE_TABLE,
      type: table.variable_name ? "dataframe" : "table",
      apply: `@${tableName}`,
      section: table.variable_name ? "Dataframe" : "Table",
      info: () => this.createTableInfoElement(tableName, table),
    };
  }

  private createTableInfoElement(
    tableName: string,
    table: DataTable,
  ): HTMLElement {
    const infoContainer = document.createElement("div");
    infoContainer.classList.add(
      "mo-cm-tooltip",
      "docs-documentation",
      "min-w-[200px]",
      "flex",
      "flex-col",
      "gap-1",
      "p-2",
    );

    // Table header with name and source
    const headerDiv = document.createElement("div");
    headerDiv.classList.add("flex", "flex-col", "gap-1");

    const nameDiv = document.createElement("div");
    nameDiv.classList.add("font-bold", "text-base");
    nameDiv.textContent = tableName;
    headerDiv.append(nameDiv);

    if (table.source) {
      const sourceDiv = document.createElement("div");
      sourceDiv.classList.add("text-xs", "text-muted-foreground");
      sourceDiv.textContent = `Source: ${table.source}`;
      headerDiv.append(sourceDiv);
    }

    infoContainer.append(headerDiv);

    // Table shape info
    const shape = [
      table.num_rows == null ? undefined : `${table.num_rows} rows`,
      table.num_columns == null ? undefined : `${table.num_columns} columns`,
    ]
      .filter(Boolean)
      .join(", ");

    if (shape) {
      const shapeDiv = document.createElement("div");
      shapeDiv.classList.add("text-xs", "bg-muted", "px-2", "py-1", "rounded");
      shapeDiv.textContent = shape;
      infoContainer.append(shapeDiv);
    }

    // Columns table
    if (table.columns && table.columns.length > 0) {
      const columnsDiv = document.createElement("div");
      columnsDiv.classList.add("overflow-auto", "max-h-60");

      const columnsTable = document.createElement("table");
      columnsTable.classList.add("w-full", "text-xs", "border-collapse");

      // Table header
      const headerRow = columnsTable.insertRow();
      headerRow.classList.add("bg-muted");

      const nameHeader = headerRow.insertCell();
      nameHeader.classList.add("p-2", "font-medium", "text-left");
      nameHeader.textContent = "Column";

      const typeHeader = headerRow.insertCell();
      typeHeader.classList.add("p-2", "font-medium", "text-left");
      typeHeader.textContent = "Type";

      // Check if any column has metadata before adding the header
      const hasAnyMetadata = table.columns.some(
        (column) => this.getItemMetadata(table, column) !== undefined,
      );

      let metadataHeader: HTMLTableCellElement | undefined;
      if (hasAnyMetadata) {
        metadataHeader = headerRow.insertCell();
        metadataHeader.classList.add("p-2", "font-medium", "text-left");
        metadataHeader.textContent = "Metadata";
      }

      // Table rows
      table.columns.forEach((column, index) => {
        const row = columnsTable.insertRow();
        row.classList.add(index % 2 === 0 ? "bg-background" : "bg-muted/30");

        const nameCell = row.insertCell();
        nameCell.classList.add("p-2", "font-mono");
        nameCell.textContent = column.name;

        const typeCell = row.insertCell();
        typeCell.classList.add("p-2", "text-muted-foreground");
        typeCell.textContent = column.type;

        if (hasAnyMetadata) {
          const metadataCell = row.insertCell();
          metadataCell.classList.add("p-2");

          const itemMetadata = this.getItemMetadata(table, column);
          if (itemMetadata) {
            metadataCell.append(itemMetadata);
          }
        }
      });

      columnsDiv.append(columnsTable);
      infoContainer.append(columnsDiv);
    }

    return infoContainer;
  }

  private getItemMetadata(
    table: DataTable,
    column: DataTable["columns"][0],
  ): HTMLSpanElement | undefined {
    const isPrimaryKey = table.primary_keys?.includes(column.name);
    const isIndexed = table.indexes?.includes(column.name);
    if (isPrimaryKey || isIndexed) {
      const badge = document.createElement("span");
      badge.textContent = isPrimaryKey ? "PK" : "IDX";
      const color = isPrimaryKey
        ? "bg-blue-100 text-blue-800 dark:bg-blue-900 dark:text-blue-200"
        : "bg-amber-100 text-amber-800 dark:bg-amber-900 dark:text-amber-200";
      badge.classList.add(
        "text-xs",
        "px-1.5",
        "py-0.5",
        "rounded-full",
        "font-medium",
        ...color.split(" "),
      );
      return badge;
    }
  }
}
