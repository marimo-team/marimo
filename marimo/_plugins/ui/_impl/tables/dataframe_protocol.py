# Copyright 2024 Marimo. All rights reserved.

# Copied from https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py
# since this is not published on PyPI.

from __future__ import annotations

import enum
from abc import (
    ABC,
    abstractmethod,
)
from typing import (
    Any,
    Dict,
    Iterable,
    Optional,
    Sequence,
    Tuple,
    TypedDict,
)


class DlpackDeviceType(enum.IntEnum):
    """Integer enum for device type codes matching DLPack."""

    CPU = 1
    CUDA = 2
    CPU_PINNED = 3
    OPENCL = 4
    VULKAN = 7
    METAL = 8
    VPI = 9
    ROCM = 10


class DtypeKind(enum.IntEnum):
    """
    Integer enum for data types.

    Attributes
    ----------
    INT : int
        Matches to signed integer data type.
    UINT : int
        Matches to unsigned integer data type.
    FLOAT : int
        Matches to floating point data type.
    BOOL : int
        Matches to boolean data type.
    STRING : int
        Matches to string data type (UTF-8 encoded).
    DATETIME : int
        Matches to datetime data type.
    CATEGORICAL : int
        Matches to categorical data type.
    """

    INT = 0
    UINT = 1
    FLOAT = 2
    BOOL = 20
    STRING = 21  # UTF-8
    DATETIME = 22
    CATEGORICAL = 23


Dtype = Tuple[DtypeKind, int, str, str]  # see Column.dtype


class CategoricalDescription(TypedDict):
    # whether the ordering of dictionary indices is semantically meaningful
    is_ordered: bool
    # whether a dictionary-style mapping of categorical values to other objects exists # noqa: E501
    is_dictionary: bool
    # Python-level only (e.g. ``{int: str}``).
    # None if not a dictionary-style categorical.
    categories: Optional[Column]


class Column(ABC):
    """
    A column object, with only the methods and properties required by the
    interchange protocol defined.

    A column can contain one or more chunks. Each chunk can contain up to three
    buffers - a data buffer, a mask buffer (depending on null representation),
    and an offsets buffer (if variable-size binary; e.g., variable-length
    strings).

    TBD: there's also the "chunk" concept here, which is implicit in Arrow as
         multiple buffers per array (= column here). Semantically it may make
         sense to have both: chunks were meant for example for lazy evaluation
         of data which doesn't fit in memory, while multiple buffers per column
         could also come from doing a selection operation on a single
         contiguous buffer.

         Given these concepts, one would expect chunks to be all of the same
         size (say a 10,000 row dataframe could have 10 chunks of 1,000 rows),
         while multiple buffers could have data-dependent lengths. Not an issue
         in pandas if one column is backed by a single NumPy array, but in
         Arrow it seems possible.
         Are multiple chunks *and* multiple buffers per column necessary for
         the purposes of this interchange protocol, or must producers either
         reuse the chunk concept for this or copy the data?

    Note: this Column object can only be produced by ``__dataframe__``, so
          doesn't need its own version or ``__column__`` protocol.
    """

    @abstractmethod
    def size(self) -> int:
        """
        Size of the column, in elements.

        Corresponds to DataFrame.num_rows() if column is a single chunk;
        equal to size of this current chunk otherwise.

        Is a method rather than a property because it may cause a (potentially
        expensive) computation for some dataframe implementations.
        """
        pass

    @property
    @abstractmethod
    def offset(self) -> int:
        """
        Offset of first element.

        May be > 0 if using chunks; for example for a column with N chunks of
        equal size M (only the last chunk may be shorter),
        ``offset = n * M``, ``n = 0 .. N-1``.
        """
        pass

    @property
    @abstractmethod
    def dtype(self) -> Dtype:
        """
        Dtype description as a tuple ``(kind, bit-width, format string, endianness)``.

        Bit-width : the number of bits as an integer
        Format string : data type description format string in Apache Arrow C
                        Data Interface format.
        Endianness : current only native endianness (``=``) is supported

        Notes:
            - Kind specifiers are aligned with DLPack where possible (hence the
              jump to 20, leave enough room for future extension)
            - Masks must be specified as boolean with either bit width 1 (for bit
              masks) or 8 (for byte masks).
            - Dtype width in bits was preferred over bytes
            - Endianness isn't too useful, but included now in case in the future
              we need to support non-native endianness
            - Went with Apache Arrow format strings over NumPy format strings
              because they're more complete from a dataframe perspective
            - Format strings are mostly useful for datetime specification, and
              for categoricals.
            - For categoricals, the format string describes the type of the
              categorical in the data buffer. In case of a separate encoding of
              the categorical (e.g. an integer to string mapping), this can
              be derived from ``self.describe_categorical``.
            - Data types not included: complex, Arrow-style null, binary, decimal,
              and nested (list, struct, map, union) dtypes.
        """  # noqa: E501
        pass

    @property
    @abstractmethod
    def describe_categorical(self) -> CategoricalDescription:
        """
        If the dtype is categorical, there are two options:
        - There are only values in the data buffer.
        - There is a separate non-categorical Column encoding categorical values.

        Raises TypeError if the dtype is not categorical

        Returns the dictionary with description on how to interpret the data buffer:
            - "is_ordered" : bool, whether the ordering of dictionary indices is
                             semantically meaningful.
            - "is_dictionary" : bool, whether a mapping of
                                categorical values to other objects exists
            - "categories" : Column representing the (implicit) mapping of indices to
                             category values (e.g. an array of cat1, cat2, ...).
                             None if not a dictionary-style categorical.

        TBD: are there any other in-memory representations that are needed?
        """  # noqa: E501
        pass

    @property
    @abstractmethod
    def null_count(self) -> Optional[int]:
        """
        Number of null elements, if known.

        Note: Arrow uses -1 to indicate "unknown", but None seems cleaner.
        """
        pass

    @property
    @abstractmethod
    def metadata(self) -> Dict[str, Any]:
        """
        The metadata for the column. See `DataFrame.metadata` for more details.
        """
        pass


class DataFrame(ABC):
    """
    A data frame class, with only the methods required by the interchange
    protocol defined.

    A "data frame" represents an ordered collection of named columns.
    A column's "name" must be a unique string.
    Columns may be accessed by name or by position.

    This could be a public data frame class, or an object with the methods and
    attributes defined on this DataFrame class could be returned from the
    ``__dataframe__`` method of a public data frame class in a library adhering
    to the dataframe interchange protocol specification.
    """

    version = 0  # version of the protocol

    @abstractmethod
    def __dataframe__(
        self, nan_as_null: bool = False, allow_copy: bool = True
    ) -> "DataFrame":
        """
        Construct a new exchange object, potentially changing the parameters.

        ``nan_as_null`` is a DEPRECATED keyword that should not be used. See warning
        below.
        ``allow_copy`` is a keyword that defines whether or not the library is
        allowed to make a copy of the data. For example, copying data would be
        necessary if a library supports strided buffers, given that this protocol
        specifies contiguous buffers.

        WARNING: the ``nan_as_null`` parameter will be removed from the API protocol.
        Please avoid passing it as either a positional or keyword argument. Call this
        method using ``.__dataframe__(allow_copy=...)``.
        """  # noqa: E501
        pass

    @property
    @abstractmethod
    def metadata(self) -> Dict[str, Any]:
        """
        The metadata for the data frame, as a dictionary with string keys. The
        contents of `metadata` may be anything, they are meant for a library
        to store information that it needs to, e.g., roundtrip losslessly or
        for two implementations to share data that is not (yet) part of the
        interchange protocol specification. For avoiding collisions with other
        entries, please add name the keys with the name of the library
        followed by a period and the desired name, e.g, ``pandas.indexcol``.
        """
        pass

    @abstractmethod
    def num_columns(self) -> int:
        """
        Return the number of columns in the DataFrame.
        """
        pass

    @abstractmethod
    def num_rows(self) -> Optional[int]:
        # TODO: not happy with Optional, but need to flag it may be expensive
        #       why include it if it may be None - what do we expect consumers
        #       to do here?
        """
        Return the number of rows in the DataFrame, if available.
        """
        pass

    @abstractmethod
    def num_chunks(self) -> int:
        """
        Return the number of chunks the DataFrame consists of.
        """
        pass

    @abstractmethod
    def column_names(self) -> Iterable[str]:
        """
        Return an iterator yielding the column names.
        """
        pass

    @abstractmethod
    def get_column(self, i: int) -> Column:
        """
        Return the column at the indicated position.
        """
        pass

    @abstractmethod
    def get_column_by_name(self, name: str) -> Column:
        """
        Return the column whose name is the indicated name.
        """
        pass

    @abstractmethod
    def get_columns(self) -> Iterable[Column]:
        """
        Return an iterator yielding the columns.
        """
        pass

    @abstractmethod
    def select_columns(self, indices: Sequence[int]) -> "DataFrame":
        """
        Create a new DataFrame by selecting a subset of columns by index.
        """
        pass

    @abstractmethod
    def select_columns_by_name(self, names: Sequence[str]) -> "DataFrame":
        """
        Create a new DataFrame by selecting a subset of columns by name.
        """
        pass

    @abstractmethod
    def get_chunks(
        self, n_chunks: Optional[int] = None
    ) -> Iterable["DataFrame"]:
        """
        Return an iterator yielding the chunks.

        By default (None), yields the chunks that the data is stored as by the
        producer. If given, ``n_chunks`` must be a multiple of
        ``self.num_chunks()``, meaning the producer must subdivide each chunk
        before yielding it.

        Note that the producer must ensure that all columns are chunked the
        same way.
        """
        pass
