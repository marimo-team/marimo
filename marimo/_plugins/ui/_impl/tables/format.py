# Copyright 2024 Marimo. All rights reserved.
from __future__ import annotations

from typing import Callable, Dict, List, Union

from marimo._plugins.core.web_component import JSONType

FormatMapping = Dict[str, Union[str, Callable[..., JSONType]]]


def format_value(
    col: str, value: JSONType, format_mapping: FormatMapping
) -> JSONType:
    # Return None if the format mapping is None
    if format_mapping is None:
        return value
    # Return None if the value is None
    if value is None:
        return None
    # Apply formatting logic based on column and value
    if col in format_mapping:
        formatter = format_mapping[col]
        if isinstance(formatter, str):
            return formatter.format(value)
        if callable(formatter):
            return formatter(value)
    return value


def format_row(
    row: Dict[str, JSONType], format_mapping: FormatMapping
) -> Dict[str, JSONType]:
    # Return None if the format mapping is None
    if format_mapping is None:
        return row
    # Apply formatting to each value in a row dictionary
    return {
        col: format_value(col, value, format_mapping)
        for col, value in row.items()
    }


def format_column(
    col: str, values: List[JSONType], format_mapping: FormatMapping
) -> List[JSONType]:
    # Return None if the format mapping is None
    if format_mapping is None:
        return values
    # Apply formatting to each value in a column list
    return [format_value(col, value, format_mapping) for value in values]
