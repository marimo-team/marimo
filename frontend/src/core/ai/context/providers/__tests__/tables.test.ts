/* Copyright 2024 Marimo. All rights reserved. */

import { describe, expect, it } from "vitest";
import type { DatasetTablesMap } from "@/core/datasets/data-source-connections";
import type { DataTable, DataTableColumn } from "@/core/kernel/messages";
import { type TableContextItem, TableContextProvider } from "../tables";

// Mock data for testing
const createMockColumn = (
  name: string,
  type: string,
  externalType?: string,
): DataTableColumn => ({
  name,
  type: type as DataTableColumn["type"],
  external_type: externalType || type,
  sample_values: [`sample_${name}_1`, `sample_${name}_2`],
});

const createMockTable = (
  name: string,
  options: Partial<DataTable> = {},
): DataTable => ({
  name,
  source: "memory",
  source_type: "local",
  type: "table",
  columns: [
    createMockColumn("id", "integer", "int64"),
    createMockColumn("name", "string", "varchar"),
    createMockColumn("active", "boolean", "bool"),
  ],
  num_rows: 100,
  num_columns: 3,
  engine: null,
  indexes: null,
  primary_keys: ["id"],
  variable_name: null,
  ...options,
});

describe("TableContextProvider", () => {
  describe("getItems", () => {
    it("should return empty array when no tables", () => {
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new TableContextProvider(tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("empty-tables");
    });

    it("should return table items for single table", () => {
      const table = createMockTable("users");
      const tablesMap: DatasetTablesMap = new Map([["users", table]]);
      const provider = new TableContextProvider(tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("single-table");
    });

    it("should return table items for multiple tables", () => {
      const usersTable = createMockTable("users");
      const ordersTable = createMockTable("orders", {
        source: "database.db",
        source_type: "duckdb",
        type: "view",
        num_rows: 250,
        num_columns: 5,
        columns: [
          createMockColumn("order_id", "integer"),
          createMockColumn("user_id", "integer"),
          createMockColumn("amount", "number", "decimal"),
          createMockColumn("created_at", "string", "timestamp"),
          createMockColumn("status", "string", "enum"),
        ],
      });
      const tablesMap: DatasetTablesMap = new Map([
        ["users", usersTable],
        ["orders", ordersTable],
      ]);
      const provider = new TableContextProvider(tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("multiple-tables");
    });

    it("should handle dataframe tables with variable names", () => {
      const dfTable = createMockTable("df_analysis", {
        variable_name: "df_analysis",
        source: "pandas",
        source_type: "local",
        columns: [
          createMockColumn("timestamp", "string", "datetime64[ns]"),
          createMockColumn("value", "number", "float64"),
          createMockColumn("category", "string", "object"),
        ],
      });
      const tablesMap: DatasetTablesMap = new Map([["df_analysis", dfTable]]);
      const provider = new TableContextProvider(tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("dataframe-table");
    });

    it("should handle tables with minimal information", () => {
      const minimalTable = createMockTable("minimal", {
        columns: [],
        num_rows: null,
        num_columns: null,
        source: "",
        primary_keys: null,
        indexes: null,
      });
      const tablesMap: DatasetTablesMap = new Map([["minimal", minimalTable]]);
      const provider = new TableContextProvider(tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("minimal-table");
    });
  });

  describe("formatContext", () => {
    it("should format context for basic table", () => {
      const table = createMockTable("products");
      const item: TableContextItem = {
        type: "data",
        uri: "products",
        name: "products",
        description: "memory",
        data: table,
      };

      const tablesMap: DatasetTablesMap = new Map([["products", table]]);
      const provider = new TableContextProvider(tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("basic-table-context");
    });

    it("should format context for table without shape info", () => {
      const table = createMockTable("no_shape", {
        num_rows: null,
        num_columns: null,
      });
      const item: TableContextItem = {
        type: "data",
        uri: "no_shape",
        name: "no_shape",
        description: "memory",
        data: table,
      };

      const tablesMap: DatasetTablesMap = new Map([["no_shape", table]]);
      const provider = new TableContextProvider(tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("no-shape-table-context");
    });

    it("should format context for table without columns", () => {
      const table = createMockTable("no_columns", {
        columns: [],
      });
      const item: TableContextItem = {
        type: "data",
        uri: "no_columns",
        name: "no_columns",
        description: "memory",
        data: table,
      };

      const tablesMap: DatasetTablesMap = new Map([["no_columns", table]]);
      const provider = new TableContextProvider(tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("no-columns-table-context");
    });

    it("should format context for remote database table", () => {
      const table = createMockTable("remote_table", {
        source: "postgresql://localhost:5432/mydb",
        source_type: "connection",
        engine: "postgresql",
        columns: [
          createMockColumn("uuid", "string", "uuid"),
          createMockColumn("created_at", "string", "timestamp with time zone"),
          createMockColumn("metadata", "string", "jsonb"),
        ],
      });
      const item: TableContextItem = {
        type: "data",
        uri: "remote_table",
        name: "remote_table",
        description: "postgresql://localhost:5432/mydb",
        data: table,
      };

      const tablesMap: DatasetTablesMap = new Map([["remote_table", table]]);
      const provider = new TableContextProvider(tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("remote-table-context");
    });
  });

  describe("provider properties", () => {
    it("should have correct provider properties", () => {
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new TableContextProvider(tablesMap);

      expect(provider.title).toBe("Tables");
      expect(provider.mentionPrefix).toBe("@");
      expect(provider.contextType).toBe("data");
    });
  });

  describe("createTableInfoElement", () => {
    it("should create info element for table with all properties", () => {
      const table = createMockTable("full_table", {
        source: "my_database.db",
        num_rows: 1000,
        num_columns: 5,
        columns: [
          createMockColumn("id", "integer", "bigint"),
          createMockColumn("name", "string", "varchar(255)"),
          createMockColumn("score", "number", "decimal(10,2)"),
        ],
      });
      const tablesMap: DatasetTablesMap = new Map([["full_table", table]]);
      const provider = new TableContextProvider(tablesMap);

      // @ts-expect-error - accessing private method for testing
      const element = provider.createTableInfoElement("full_table", table);

      expect(element.tagName).toBe("DIV");
      expect(element.classList.contains("mo-cm-tooltip")).toBe(true);
      expect(element.textContent).toMatchInlineSnapshot(
        `"full_tableSource: my_database.db1000 rows, 5 columnsColumnTypeMetadataidintegerPKnamestringscorenumber"`,
      );
    });

    it("should create info element for table without source", () => {
      const table = createMockTable("no_source", {
        source: "",
      });
      const tablesMap: DatasetTablesMap = new Map([["no_source", table]]);
      const provider = new TableContextProvider(tablesMap);

      // @ts-expect-error - accessing private method for testing
      const element = provider.createTableInfoElement("no_source", table);

      expect(element.textContent).toContain("no_source");
      expect(element.textContent).not.toContain("Source:");
    });

    it("should create info element for table without columns", () => {
      const table = createMockTable("no_cols", {
        columns: [],
      });
      const tablesMap: DatasetTablesMap = new Map([["no_cols", table]]);
      const provider = new TableContextProvider(tablesMap);

      // @ts-expect-error - accessing private method for testing
      const element = provider.createTableInfoElement("no_cols", table);

      expect(element.textContent).toContain("no_cols");
      // Should not contain a table element
      expect(element.querySelector("table")).toBeNull();
    });
  });
});
