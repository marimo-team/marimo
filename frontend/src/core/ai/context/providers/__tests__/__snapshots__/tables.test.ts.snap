// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableContextProvider > formatContext > should format context for basic table > basic-table-context 1`] = `
"Table: products
Source: memory
Shape: 100 rows, 3 columns
Columns:
  - id: integer
  - name: string
  - active: boolean"
`;

exports[`TableContextProvider > formatContext > should format context for remote database table > remote-table-context 1`] = `
"Table: remote_table
Source: postgresql://localhost:5432/mydb
Shape: 100 rows, 3 columns
Columns:
  - uuid: string
  - created_at: string
  - metadata: string"
`;

exports[`TableContextProvider > formatContext > should format context for table without columns > no-columns-table-context 1`] = `
"Table: no_columns
Source: memory
Shape: 100 rows, 3 columns"
`;

exports[`TableContextProvider > formatContext > should format context for table without shape info > no-shape-table-context 1`] = `
"Table: no_shape
Source: memory
Columns:
  - id: integer
  - name: string
  - active: boolean"
`;

exports[`TableContextProvider > getItems > should handle dataframe tables with variable names > dataframe-table 1`] = `
[
  {
    "data": {
      "columns": [
        {
          "external_type": "datetime64[ns]",
          "name": "timestamp",
          "sample_values": [
            "sample_timestamp_1",
            "sample_timestamp_2",
          ],
          "type": "string",
        },
        {
          "external_type": "float64",
          "name": "value",
          "sample_values": [
            "sample_value_1",
            "sample_value_2",
          ],
          "type": "number",
        },
        {
          "external_type": "object",
          "name": "category",
          "sample_values": [
            "sample_category_1",
            "sample_category_2",
          ],
          "type": "string",
        },
      ],
      "engine": null,
      "indexes": null,
      "name": "df_analysis",
      "num_columns": 3,
      "num_rows": 100,
      "primary_keys": [
        "id",
      ],
      "source": "pandas",
      "source_type": "local",
      "type": "table",
      "variable_name": "df_analysis",
    },
    "description": "pandas",
    "name": "df_analysis",
    "type": "data",
    "uri": "data://df_analysis",
  },
]
`;

exports[`TableContextProvider > getItems > should handle tables with minimal information > minimal-table 1`] = `
[
  {
    "data": {
      "columns": [],
      "engine": null,
      "indexes": null,
      "name": "minimal",
      "num_columns": null,
      "num_rows": null,
      "primary_keys": null,
      "source": "",
      "source_type": "local",
      "type": "table",
      "variable_name": null,
    },
    "description": "",
    "name": "minimal",
    "type": "data",
    "uri": "data://minimal",
  },
]
`;

exports[`TableContextProvider > getItems > should return empty array when no tables > empty-tables 1`] = `[]`;

exports[`TableContextProvider > getItems > should return table items for multiple tables > multiple-tables 1`] = `
[
  {
    "data": {
      "columns": [
        {
          "external_type": "int64",
          "name": "id",
          "sample_values": [
            "sample_id_1",
            "sample_id_2",
          ],
          "type": "integer",
        },
        {
          "external_type": "varchar",
          "name": "name",
          "sample_values": [
            "sample_name_1",
            "sample_name_2",
          ],
          "type": "string",
        },
        {
          "external_type": "bool",
          "name": "active",
          "sample_values": [
            "sample_active_1",
            "sample_active_2",
          ],
          "type": "boolean",
        },
      ],
      "engine": null,
      "indexes": null,
      "name": "users",
      "num_columns": 3,
      "num_rows": 100,
      "primary_keys": [
        "id",
      ],
      "source": "memory",
      "source_type": "local",
      "type": "table",
      "variable_name": null,
    },
    "description": "in-memory",
    "name": "users",
    "type": "data",
    "uri": "data://users",
  },
  {
    "data": {
      "columns": [
        {
          "external_type": "integer",
          "name": "order_id",
          "sample_values": [
            "sample_order_id_1",
            "sample_order_id_2",
          ],
          "type": "integer",
        },
        {
          "external_type": "integer",
          "name": "user_id",
          "sample_values": [
            "sample_user_id_1",
            "sample_user_id_2",
          ],
          "type": "integer",
        },
        {
          "external_type": "decimal",
          "name": "amount",
          "sample_values": [
            "sample_amount_1",
            "sample_amount_2",
          ],
          "type": "number",
        },
        {
          "external_type": "timestamp",
          "name": "created_at",
          "sample_values": [
            "sample_created_at_1",
            "sample_created_at_2",
          ],
          "type": "string",
        },
        {
          "external_type": "enum",
          "name": "status",
          "sample_values": [
            "sample_status_1",
            "sample_status_2",
          ],
          "type": "string",
        },
      ],
      "engine": null,
      "indexes": null,
      "name": "orders",
      "num_columns": 5,
      "num_rows": 250,
      "primary_keys": [
        "id",
      ],
      "source": "database.db",
      "source_type": "duckdb",
      "type": "view",
      "variable_name": null,
    },
    "description": "database.db",
    "name": "orders",
    "type": "data",
    "uri": "data://orders",
  },
]
`;

exports[`TableContextProvider > getItems > should return table items for single table > single-table 1`] = `
[
  {
    "data": {
      "columns": [
        {
          "external_type": "int64",
          "name": "id",
          "sample_values": [
            "sample_id_1",
            "sample_id_2",
          ],
          "type": "integer",
        },
        {
          "external_type": "varchar",
          "name": "name",
          "sample_values": [
            "sample_name_1",
            "sample_name_2",
          ],
          "type": "string",
        },
        {
          "external_type": "bool",
          "name": "active",
          "sample_values": [
            "sample_active_1",
            "sample_active_2",
          ],
          "type": "boolean",
        },
      ],
      "engine": null,
      "indexes": null,
      "name": "users",
      "num_columns": 3,
      "num_rows": 100,
      "primary_keys": [
        "id",
      ],
      "source": "memory",
      "source_type": "local",
      "type": "table",
      "variable_name": null,
    },
    "description": "in-memory",
    "name": "users",
    "type": "data",
    "uri": "data://users",
  },
]
`;
